### PR TITLE
Melina97/adjoint

### PR DESCRIPTION
**Pull request by Melina97**
For time-dependent problems we have previously recomputed the symbolic derivation of the adjoint at every time step, which can dominate the run time of small adjoint problems. Since the symbolic representation of dFdu* remains unchanged, it suffices to compute it once and exchange the numerical value of each coefficient with its checkpointed value at every time step. 

We can’t do the same for dFdm* because we don’t know m beforehand. To reuse the symbolic representation of the adjoint for the same m, we build a cache instead.

### CODE REVIEW

File: firedrake/adjoint/blocks.py [hunk 1]
```diff
@@ -146,6 +148,83 @@ def _ad_nlvs_replace_forms(self):
...
+
+        if isinstance(c, firedrake.Function):
+            trial_function = firedrake.TrialFunction(c.function_space())
+        elif isinstance(c, firedrake.Constant):
+            mesh = self.compat.extract_mesh_from_form(F_form)
+            trial_function = firedrake.TrialFunction(c._ad_function_space(mesh))
+        elif isinstance(c, self.compat.ExpressionType):
+            mesh = F_form.ufl_domain().ufl_cargo()
+            c_fs = c._ad_function_space(mesh)
+            trial_function = firedrake.TrialFunction(c_fs)
```
Comment (stephankramer) [inline firedrake/adjoint/blocks.py:194]: I _think_ this can probably go given #2202

File: firedrake/adjoint/blocks.py [hunk 2]
```diff
@@ -146,6 +148,83 @@ def _ad_nlvs_replace_forms(self):
...
+            self._dFdm_cache[c] = dFdm
+
+        # Replace the form coefficients with checkpointed values.
+        replace_map = self._replace_map(dFdm)
+        dFdm = replace(dFdm, replace_map)
+
+        dFdm = dFdm * adj_sol
+        dFdm = self.compat.assemble_adjoint_value(dFdm, **self.assemble_kwargs)
+
+        if isinstance(c, self.compat.ExpressionType):
```
Comment (stephankramer) [inline firedrake/adjoint/blocks.py:223]: Same here

File: firedrake/assemble_expressions.py [hunk 3]
```diff
@@ -457,12 +457,13 @@ def __exit__(self, *args, **kwargs):
 
 @PETSc.Log.EventDecorator()
 @known_pyop2_safe
-def evaluate_expression(expr, subset=None):
+def evaluate_expression(expr, subset=None, dat_map=None):
```
Comment (stephankramer) [inline firedrake/assemble_expressions.py:460]: What is this for?

**Review by stephankramer — COMMENTED** [review]
All looks good to me.
This is very useful in fact. Below is a profile from an adjoint calculation in an actual real-life application (a Navier Stokes model of ocean flow under and ice shelf using https://github.com/thwaitesproject/thwaites/). As you can see it was actually spending 110 out of 280s merely taking the (symbolic) adjoint!
```
In [13]: p.sort_stats(SortKey.CUMULATIVE).print_stats(20)
Wed Aug  5 12:46:37 2020    stats

         672171905 function calls (659920149 primitive calls) in 284.406 seconds

   Ordered by: cumulative time
   List reduced from 22299 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   6299/1    0.024    0.000  284.462  284.462 {built-in method builtins.exec}
        1    0.009    0.009  284.462  284.462 ice_shelf_coarse.py:5(<module>)
5024/4410    0.025    0.000  228.719    0.052 tape.py:43(wrapper)
        1    0.002    0.002  227.137  227.137 tape.py:138(evaluate_adj)
      598    0.033    0.000  227.122    0.380 block.py:97(evaluate_adj)
     2088    0.092    0.000  179.219    0.086 solving.py:181(evaluate_adj_component)
152066/23755    0.321    0.000  138.514    0.006 map_integrands.py:22(map_integrands)
    21835    0.034    0.000  137.234    0.006 map_integrands.py:45(map_integrand_dags)
184847/129303   16.197    0.000  129.945    0.001 map_dag.py:30(map_expr_dags)
182372/128936    0.557    0.000  126.663    0.001 map_dag.py:17(map_expr_dag)
   126121    0.073    0.000  125.788    0.001 map_integrands.py:46(<lambda>)
    14977    0.109    0.000  124.443    0.008 map_integrands.py:27(<listcomp>)
     1920    0.075    0.000  110.082    0.057 ufl_expr.py:200(adjoint)
     1920    0.014    0.000  105.090    0.055 formoperators.py:122(adjoint)
     4472    0.125    0.000  102.821    0.023 ad.py:17(expand_derivatives)
     7676    0.019    0.000   69.304    0.009 apply_derivatives.py:1088(apply_derivatives)
    10572    0.067    0.000   63.736    0.006 assemble.py:630(_assemble)
     2032    0.019    0.000   61.502    0.030 assembly.py:8(wrapper)
     2032    0.045    0.000   61.426    0.030 assemble.py:29(assemble)
     8075    0.244    0.000   60.961    0.008 assemble.py:493(create_parloops)
```

File: firedrake/assemble_expressions.py (hunk 3 above)
Comment (Melina97) [reply firedrake/assemble_expressions.py:460]: That was relevant to a different part of my project, I'll remove it here. Thanks!

File: firedrake/adjoint/blocks.py [hunk 4]
```diff
@@ -146,6 +148,76 @@ def _ad_nlvs_replace_forms(self):
         self._ad_assign_coefficients(problem.F)
         self._ad_assign_coefficients(problem.J)
 
+    def prepare_evaluate_adj(self, inputs, adj_inputs, relevant_dependencies):
+        dJdu = adj_inputs[0]
+
+        F_form = self._create_F_form()
+
+        dFdu_form = self.adj_F
+        dJdu = dJdu.copy()
```
Comment (sv2518) [inline firedrake/adjoint/blocks.py:157]: What is `adj_inputs`? Could it be that you need a deepcopy here?

Comment (stephankramer) [reply firedrake/adjoint/blocks.py:157]: `adj_inputs` are the adjoint values associated with the output variables (which become inputs as we traverse backwards through the tree in the adjoint). Not entirely sure why a copy is needed here but it's the same it was doing previously in the prepare_evaluate_adj of GenericSolveBlock that is being overloaded here.

**Review by dham — APPROVED** [review]

**Pull request merged**